File: ProjectEuler/Problems/Problem024.cs
using Microsoft.Extensions.Logging;

namespace ProjectEuler.Problems
{
    public class Problem024 : Problem<string>
    {
        public Problem024(ILogger<Problem024> logger) : base(logger) { }

        protected override string CalculateAnswer()
        {
            var characters = "0123456789";
            var elements = characters.ToCharArray();
            var n = 1000000;

            var index = 1;
            var returnedToFirstPermutation = false;
            while (true)
            {
                if (index % 10000 == 0) Logger.LogDebug("{index}: {elements}", index, new string(elements));
                if (index == n || returnedToFirstPermutation)
                {
                    break;
                }
                var result = MathHelpers.LexicographicPermute(elements);
                ++index;
                if (result) returnedToFirstPermutation = true;
            }

            var answer = new string(elements);
            return answer;
        }
    }
}

File: ProjectEulerPython/problems/problem_024.py
import math_helpers
from problems.problem import Problem

class Problem024(Problem):

  def calculate_answer(self) -> int:
    elements = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    n = 1000000
    batch = n // 20

    index = 1
    returned_first_permutation = False
    while True:
      if index % batch == 0:
        self.print_detail(str(index) + ": " + str(elements))
      if index == n or returned_first_permutation:
        break
      result = math_helpers.lexicographic_permute(elements)
      index += 1
      if result:
        returned_first_permutation = True

    answer = ""
    for i in range(0, len(elements)):
      answer += str(elements[i])
    return answer